Sources/swift-parser-test/DiagnosticsFormatter.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  File.swift
+//  
+//
+//  Created by Felix Wehnert on 02.10.22.
+//
+
+import Foundation
+import SwiftParser
+import SwiftSyntax
+import SwiftDiagnostics
+
+
+// To put `Diagnostic`s into a set they must conform to hashable.
+// Two `Diagnostic` instances are the same, if they contain the same diagnosticID at the same node.
+extension Diagnostic: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(self.node)
+  }
+  
+  public static func ==(lhs: Diagnostic, rhs: Diagnostic) -> Bool {
+    return lhs.node == rhs.node && lhs.diagnosticID == rhs.diagnosticID
+  }
+}
+
+struct DiagnosticsFormatter {
+  
+  /**
+   A wrapper struct for a source line and its diagnostics
+   */
+  private struct SourceLine {
+    var diagnostics = Set<Diagnostic>()
+    var sourceString: String
+  }
+  
+  /**
+   Number of lines which should be printed before and after the diagnostic message
+   */
+  static let contextSize = 2
+  
+  /**
+   Print given diagnostics for a given syntax tree on the command line
+   */
+  static func printDiagnostics(tree: SourceFileSyntax, diags: [Diagnostic]) {
+    
+    let slc = SourceLocationConverter(file: "", tree: tree)
+    
+    // First, we need to put each line and its diagnostics together
+    var sourceLines = [SourceLine]()
+    
+    for sourceLine in tree.description.split(separator: "\n", omittingEmptySubsequences: false) {
```

Comment:
@ahoppen Is there a better way of getting each source code line?

Reply:
`SourceLocationConverter` deals with the lines of a source file, perhaps you can start with that and extend it with whatever utility methods you need?

Reply:
Here’s a diff to get the source locations from `SourceLocationConverter`:

```diff
diff --git a/Sources/SwiftSyntax/SourceLocation.swift b/Sources/SwiftSyntax/SourceLocation.swift
index d37035cc..b3da7967 100644
--- a/Sources/SwiftSyntax/SourceLocation.swift
+++ b/Sources/SwiftSyntax/SourceLocation.swift
@@ -114,6 +114,8 @@ public struct SourceRange: Hashable, Codable, CustomDebugStringConvertible {
 /// part of the same tree that was used to initialize this class.
 public final class SourceLocationConverter {
   let file: String
+  /// The source of the file, modelled as data so it can contain invalid UTF-8.
+  let source: [UInt8]
   /// Array of lines and the position at the start of the line.
   let lines: [AbsolutePosition]
   /// Position at end of file.
@@ -125,6 +127,7 @@ public final class SourceLocationConverter {
   public init<SyntaxType: SyntaxProtocol>(file: String, tree: SyntaxType) {
     assert(tree.parent == nil, "SourceLocationConverter must be passed the root of the syntax tree")
     self.file = file
+    self.source = tree.syntaxTextBytes
     (self.lines, endOfFile) = computeLines(tree: Syntax(tree))
     assert(tree.byteSize == endOfFile.utf8Offset)
   }
@@ -134,10 +137,35 @@ public final class SourceLocationConverter {
   ///   - source: The source code to convert positions to line/columns for.
   public init(file: String, source: String) {
     self.file = file
+    self.source = Array(source.utf8)
     (self.lines, endOfFile) = computeLines(source)
     assert(source.utf8.count == endOfFile.utf8Offset)
   }
 
+  /// Execute the body with an array that contains each source line.
+  func withSourceLines<T>(_ body: ([SyntaxText]) throws -> T) rethrows -> T {
+    return try source.withUnsafeBufferPointer { (sourcePointer: UnsafeBufferPointer<UInt8>) in
+      var result: [SyntaxText] = []
+      var previousLoc = AbsolutePosition.startOfFile
+      for lineStartLoc in lines.dropFirst() + [endOfFile] {
+        result.append(SyntaxText(
+          baseAddress: sourcePointer.baseAddress?.advanced(by: previousLoc.utf8Offset),
+          count: lineStartLoc.utf8Offset - previousLoc.utf8Offset
+        ))
+        previousLoc = lineStartLoc
+      }
+      return try body(result)
+    }
+  }
+
+  /// Return the source lines of the file as `String`s.
+  /// Because `String` cannot model invalid UTF-8, the concatenation of these source lines might not be source-accurate in case there are Unicode errors in the source file, but for most practical purposes, this should not pose an issue.
+  public var sourceLines: [String] {
+    return withSourceLines { syntaxTextLines in
+      return syntaxTextLines.map { String(syntaxText: $0) }
+    }
+  }
+
   /// Convert a `AbsolutePosition` to a `SourceLocation`. If the position is
   /// exceeding the file length then the `SourceLocation` for the end of file
   /// is returned. If position is negative the location for start of file is
diff --git a/Sources/swift-parser-test/DiagnosticsFormatter.swift b/Sources/swift-parser-test/DiagnosticsFormatter.swift
index 1d825288..27c3f647 100644
--- a/Sources/swift-parser-test/DiagnosticsFormatter.swift
+++ b/Sources/swift-parser-test/DiagnosticsFormatter.swift
@@ -54,7 +54,7 @@ struct DiagnosticsFormatter {
     // First, we need to put each line and its diagnostics together
     var sourceLines = [SourceLine]()
     
-    for sourceLine in tree.description.split(separator: "\n", omittingEmptySubsequences: false) {
+    for sourceLine in slc.sourceLines {
       
       let diagsForLine = diags.filter { diag in
         let line = diag.location(converter: slc).line

```